### PR TITLE
Automated Performance Monitoring (CI Integration)

### PR DESCRIPTION
**Pull request by kotsaloscv**
Automated Performance Monitoring (CI Integration)

### CODE REVIEW

**Comment by cattabiani** [comment]
do we need the `monitoring` in the name?

**Comment by kotsaloscv** [comment]
> do we need the `monitoring` in the name?

I think yes :)

**Comment by cattabiani** [comment]
ehm... why?

**Comment by kotsaloscv** [comment]
because we monitor the performance by comparing the latest STEPS version with some reference commit.

**Comment by cattabiani** [comment]
Isn't it already clear from the fact that it is in the CI? What is not `monitoring` but in the CI?

File: SimpleModel/profiling/caliper.batch [hunk 1]
```diff
@@ -1,21 +1,17 @@
...
-#SBATCH --cpus-per-task=2
-#SBATCH --exclusive
-#SBATCH --mem=0
-#SBATCH --output="%x.out"
+module load unstable python-dev python
 
-module purge
-module load archive/2021-12
-module load python-dev
+source ./python-venv/bin/activate
```
Comment (tristan0x) [inline SimpleModel/profiling/caliper.batch:5]: This virtual environment is not mentioned in the documentation. If you really need to install additional Python packages, this sbatch should create it if missing.

```bash
[ -d python-venv ] || python3 -mvenv python-venv
source ./python-venv/bin/activate
pip install ...
```

File: SimpleModel/profiling/caliper.batch [hunk 2]
```diff
@@ -1,21 +1,17 @@
...
 
-# spack dev-build steps@develop+distmesh+petsc
-export PYTHONPATH="$STEPS_INSTALL_DIR:$PYTHONPATH"
+echo "*******************************************************************************"
+echo "SPACK_INSTALLED_HASH (steps) : " $SPACK_INSTALLED_HASH
```
Comment (tristan0x) [inline SimpleModel/profiling/caliper.batch:8]: This script should abort if:
* the `SPACK_INSTALLED_HASH` environment variable is not set.
* `spack find --paths /$SPACK_INSTALLED_HASH` cannot find the hash


I agree it should have been the case for` `SPACK_INSTALL_DIR` already...

File: SimpleModel/profiling/caliper.batch [hunk 3]
```diff
@@ -1,21 +1,17 @@
...
+echo "*******************************************************************************"
+export PYTHONPATH=`spack find --paths /$SPACK_INSTALLED_HASH | tail -n 1 | grep -o "/.*"`:$PYTHONPATH 
+echo "*******************************************************************************"
+echo "PYTHONPATH : " $PYTHONPATH
+echo "*******************************************************************************"
 
-spack load caliper # or module load caliper, depending on your setup
```
Comment (tristan0x) [inline SimpleModel/profiling/caliper.batch:18]: I do not understand why the `caliper` module is not need anymore.

File: caBurst/background/extra/constants.py [hunk 4]
```diff
@@ -0,0 +1,394 @@
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+#
```
Comment (cattabiani) [inline caBurst/background/extra/constants.py:2]: also, I would try to put these constants in a class or something so they are easier to import. I had this problem when I wanted to structure a little bit more my code. An example: if you want to bring the main function of your simulation in a class member you cannot do `import constants` and getting these constants is clunky. 

It is probably material for a future PR

File: profiling/README.md [hunk 5]
```diff
@@ -1,28 +1,18 @@
-Here, we describe how to perform the strong scaling & Roofline analysis found in STEPS 4 Release paper. To simply re-generate the graphs of the paper just run the `STEPS_PerfGraphs.ipynb` jupyter notebook (all data baked in).
+Here, we describe how to perform the strong scaling & Roofline analysis found in STEPS 4 Release paper. To simply re-generate the graphs of the paper just run the `STEPS_PerfGraphs.ipynb` jupyter notebook (all data is baked in).
```
Comment (cattabiani) [inline profiling/README.md:1]:
```suggestion
Here, we describe how to perform the strong scaling & Roofline analysis found in STEPS 4 Release paper. To simply re-generate the graphs of the paper just run the `STEPS_PerfGraphs.ipynb` jupyter notebook (all data is backed in).
```

File: SimpleModel/profiling/caliper.batch (hunk 2 above)
Comment (kotsaloscv) [reply SimpleModel/profiling/caliper.batch:16]: Fully agree & I have corrected it.

File: SimpleModel/profiling/caliper.batch (hunk 1 above)
Comment (kotsaloscv) [reply SimpleModel/profiling/caliper.batch:5]: The setup of the python virtual env is happening in the Jupyter notebook and this is why I am not checking at it any further.

File: rallpack3/README.md [hunk 6]
```diff
@@ -1,3 +1,30 @@
...
+
+In order to run the STEPS 3 version of the model you just need to:
+
+- Decide an rng seed. For example: 
+`export rng_seed=1`
+- Select a mesh file from the mesh folder. For example: 
+`export mesh_file=mesh/axon_cube_L1000um_D866nm_1135tets.msh`
+- Finally, run the command:
+`python rallpack1_STEPS3.py ${rng_seed} ${mesh_file}`
+- Raw traces will appear in the result folder: `raw_traces/STEPS3`
```
Comment (cattabiani) [inline rallpack3/README.md:17]:
```suggestion
`export mesh_file=mesh/axon_cube_L1000um_D866nm_1135tets.msh`
- Decide of you want to run with steps 3 or 4. For example:
`steps_version=4`
- Finally, run the command:
`python rallpack1_STEPS3.py ${rng_seed} ${mesh_file} ${steps_version}`
- Raw traces will appear in the result folder: `raw_traces/STEPS${steps_version}`
```

File: rallpack3/README.md [hunk 7]
```diff
@@ -1,3 +1,30 @@
 # Rallpack 3: validations
 
-In this model we compare STEPS 3 and STEPS 4 solutions. It is suggested to run this model on a cluster. In this folder we provide some example scripts adated to run on the bb5 cluster and locally.
\ No newline at end of file
+In this model we can compare STEPS 3 and STEPS 4 solutions. It is suggested to run 
+it on a cluster. `run.sbatch` is 
+an example of a slurm job.
+                 
+## STEPS 3
```
Comment (cattabiani) [inline rallpack3/README.md:7]:
```suggestion
```

File: SimpleModel/profiling/caliper.batch [hunk 8]
```diff
@@ -1,21 +1,17 @@
...
-# Cascade Lake nodes on BB5 (intel codename clx)
-#SBATCH --constraint=clx
-#SBATCH --time=1-00:00:00
-#SBATCH --cpus-per-task=2
-#SBATCH --exclusive
-#SBATCH --mem=0
-#SBATCH --output="%x.out"
+module load unstable python-dev python
 
-module purge
```
[the remark this answers is not in the document]
Comment (kotsaloscv) [reply SimpleModel/profiling/caliper.batch:11]: I had it in the past and I faced issues with the CI. This is why I removed it.
Any idea on how to activate it if not using the CI?

File: rallpack3/README.md [hunk 9]
```diff
@@ -1,3 +1,30 @@
...
+In order to run the STEPS 4 version of the model you just need to:
+
+- Decide an rng seed. For example: 
+`export rng_seed=1`
+- Select a mesh file from the mesh folder. For example: 
+`export mesh_file=mesh/axon_cube_L1000um_D866nm_1135tets.msh`
+- Finally, run the command:
+`python rallpack1_STEPS4.py ${rng_seed} ${mesh_file}`
+- Raw traces will appear in the result folder: `raw_traces/STEPS4`
+
```
Comment (cattabiani) [inline rallpack3/README.md:30]:
```suggestion
```

File: rallpack3/rallpack3.py [hunk 10]
```diff
@@ -18,216 +18,218 @@
...
+            OC_L = OhmicCurr.Create(Leak[leaksus], g_leak_sc, leak_rev)
+
+    # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+
+    # Create the solver objects
+    rng = RNG('r123', 512, seed)
+    #rng = RNG('mt19937', 512, seed)
+
+    if steps_version == 4:
+        sim = Simulation('DistTetOpSplit', mdl, mesh, rng)#, searchMethod=NextEventSearchMethod.GIBSON_BRUCK)
```
Comment (cattabiani) [inline rallpack3/rallpack3.py:181]:
```suggestion
        sim = Simulation('DistTetOpSplit', mdl, mesh, rng, searchMethod=NextEventSearchMethod.GIBSON_BRUCK)
```

Comment (cattabiani) [reply rallpack3/rallpack3.py:181]: this was an error that was going around for a long time

**Review by cattabiani — CHANGES_REQUESTED** [review]
I mostly changed my own stuff because I noted that it was not updated/wrong

File: SimpleModel/profiling/caliper.batch (hunk 8 above)
[the remark this answers is not in the document]
Comment (tristan0x) [reply SimpleModel/profiling/caliper.batch:11]: You can test the presence of environment variable `CI_PIPELINE_ID`, or any other variable defined by GitLab CI https://docs.gitlab.com/ee/ci/variables/predefined_variables.html

**Pull request merged**